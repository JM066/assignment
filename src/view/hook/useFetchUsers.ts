import { useEffect, useState } from 'react'
import axios, { AxiosResponse } from 'axios'

type FetchData<T> = {
    data: T | undefined
    loading: boolean
    error: Error | undefined
}
export default function useFetchUsers<T>(query: string): FetchData<T> {
    const [data, setData] = useState<T | undefined>()
    const [loading, setLoading] = useState<boolean>(false)
    const [error, setError] = useState<Error>()

    useEffect(() => {
        if (!query) return
        const controller = new AbortController()
        const fetchData = async () => {
            setLoading(true)
            try {
                const data: AxiosResponse = await axios.get(
                    `${process.env.REACT_APP_BASE_URL}/${query}`,
                    { signal: controller.signal }
                )
                setData(data?.data)
            } catch (err) {
                setError(error)
            } finally {
                setLoading(false)
            }
        }
        fetchData()
        return () => {
            controller.abort()
        }
    }, [query])
    return { data, loading, error }
}
