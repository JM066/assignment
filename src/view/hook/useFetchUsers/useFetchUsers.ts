import { useEffect, useState, useRef } from 'react'
import axios from 'axios'
import { AxiosResponse, AxiosError } from 'axios'

type FetchData<T> = {
    data: T | undefined
    loading: boolean
    error: AxiosError | null
    fetchData: () => Promise<void>
}
export default function useFetchUsers<T>(query: string): FetchData<T> {
    const [data, setData] = useState<T | undefined>()
    const [loading, setLoading] = useState<boolean>(false)
    const [error, setError] = useState<AxiosError | null>(null)

    const cancel = useRef(false)

    const fetchData = async () => {
        setLoading(true)
        try {
            const data: AxiosResponse = await axios.get(
                `https://jsonplaceholder.typicode.com/${query}`
            )
            setData(data?.data)
        } catch (err) {
            if (err instanceof AxiosError) {
                setError(err)
            } else {
                throw error
            }
        } finally {
            setLoading(false)
        }
    }
    useEffect(() => {
        if (!query) return
        if (cancel.current === false) {
            fetchData()
            return () => {
                cancel.current = true
            }
        }
    }, [query])

    return { data, loading, error, fetchData }
}
