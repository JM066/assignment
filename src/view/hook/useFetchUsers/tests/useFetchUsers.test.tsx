import { renderHook, waitFor } from '@testing-library/react'
import useFetchUsers from '../useFetchUsers'
import mockAxios from 'jest-mock-axios'

jest.mock('axios')

describe('fetchData', () => {
    it('fetches users successfully', async () => {
        const { result } = renderHook(() => useFetchUsers('users'))
        await waitFor(() => {
            expect(result.current.loading).toBeFalsy()
            expect(result.current.data).toMatchSnapshot()
        })
    })
    it('fetch fails', async () => {
        const { result } = renderHook(() => useFetchUsers('users'))
        const message = 'Network Error'
        mockAxios.get.mockRejectedValueOnce(new Error(message))
        await waitFor(() => {
            expect(result.current.data).toEqual(undefined)
            expect(result.current.loading).toBeTruthy()
        })
    })
})
