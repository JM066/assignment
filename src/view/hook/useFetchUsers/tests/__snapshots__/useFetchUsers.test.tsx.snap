// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetchData fetches users successfully 1`] = `undefined`;
